Can't start GDB under virtualenv on windows

The multiprocessing module fails to find the main module when forking for the gdb server process.

The difference between virtualenv/not seems to be:

under virtual env
```
main_path: C:\yotta\workspace\Scripts\valinor.exe\__main__.py
main_name: __main__
```

not under a virtual env:
```
main_path: C:\yotta\python\Scripts\valinor-script.py
main_name: valinor-script
```

The error message in a virtual env is:

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\yotta\python\Lib\multiprocessing\forking.py", line 380, in main
    prepare(preparation_data)
  File "C:\yotta\python\Lib\multiprocessing\forking.py", line 492, in prepare
    assert main_name not in sys.modules, main_name
AssertionError: __main__
Traceback (most recent call last):
  File "C:\yotta\python\Lib\runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "C:\yotta\python\Lib\runpy.py", line 72, in _run_code
    exec code in run_globals
  File "C:\yotta\workspace\Scripts\valinor.exe\__main__.py", line 9, in <module>
  File "c:\yotta\workspace\lib\site-packages\valinor\main.py", line 140, in main
    launch_fn(projectfiles['files'], executable)
  File "c:\yotta\workspace\lib\site-packages\valinor\gdb.py", line 90, in launch_arm_gdb
    raise Exception('gdb server failed to start')
Exception: gdb server failed to start
```

I can't find a lot of other reports of this error, which suggests either valinor is doing something wrong, or the virtualenv setup is broken somehow.

cc @sg-

--- a/.coveragerc
+++ b/.coveragerc
@@ -1,4 +1,4 @@
 [run]
-concurrency = multiprocessing
+concurrency = thread
 parallel = True
 include = ./valinor/*

--- a/valinor/gdb.py
+++ b/valinor/gdb.py
@@ -23,9 +23,6 @@ def _ignoreSignal(signum, frame):
 
 def _launchPyOCDGDBServer(msg_queue):
     logger.info('preparing PyOCD gdbserver...')
-    # ignore Ctrl-C, so we don't interrupt the GDB server when the
-    # being-debugged program is being stopped:
-    #signal.signal(signal.SIGINT, _ignoreSignal);
     from pyOCD.gdbserver import GDBServer
     from pyOCD.board import MbedBoard
     gdb = None

--- a/valinor/gdb.py
+++ b/valinor/gdb.py
@@ -6,7 +6,7 @@
 
 import subprocess
 import logging
-import multiprocessing
+import threading
 import signal
 import traceback
 try:
@@ -22,36 +22,42 @@ def _ignoreSignal(signum, frame):
     ))
 
 def _launchPyOCDGDBServer(msg_queue):
-    logger.info('starting PyOCD gdbserver...')
+    logger.info('preparing PyOCD gdbserver...')
     # ignore Ctrl-C, so we don't interrupt the GDB server when the
     # being-debugged program is being stopped:
-    signal.signal(signal.SIGINT, _ignoreSignal);
+    #signal.signal(signal.SIGINT, _ignoreSignal);
     from pyOCD.gdbserver import GDBServer
     from pyOCD.board import MbedBoard
     gdb = None
     try:
-        board_selected = MbedBoard.chooseBoard()
-        with board_selected as board:
-            gdb = GDBServer(
-                board, 3333, {
-                     'break_at_hardfault': True,
-                    'step_into_interrupt': False,
-                         'break_on_reset': False,
-                }
-            )
-            if gdb.isAlive():
-                msg_queue.put('alive')
-                while gdb.isAlive():
-                    gdb.join(timeout = 0.5)
-                    # check for a "kill" message from the parent process:
-                    try:
-                        msg = msg_queue.get(False)
-                        if msg == 'kill':
-                            gdb.stop()
-                            break
-                    except Queue.Empty:
-                        pass
+        logger.info('finding connected board...')
+        board_selected = MbedBoard.chooseBoard(blocking=False)
+        if board_selected is not None:
+            with board_selected as board:
+                logger.info('starting PyOCD gdbserver...')
+                gdb = GDBServer(
+                    board, 3333, {
+                         'break_at_hardfault': True,
+                        'step_into_interrupt': False,
+                             'break_on_reset': False,
+                    }
+                )
+                if gdb.isAlive():
+                    msg_queue.put('alive')
+                    while gdb.isAlive():
+                        gdb.join(timeout = 0.5)
+                        # check for a "kill" message from the parent process:
+                        try:
+                            msg = msg_queue.get(False)
+                            if msg == 'kill':
+                                gdb.stop()
+                                break
+                        except Queue.Empty:
+                            pass
+        else:
+            logger.error('failed to find a connected board')
     except Exception as e:
+        logger.error('exception in GDB server thread: %s', e)
         if gdb != None:
             gdb.stop()
         raise
@@ -74,22 +80,29 @@ def launcher(gdb_exe):
 def arm_none_eabi_launcher(gdb_exe):
     gdb_launcher = launcher(gdb_exe)
     def launch_arm_gdb(projectfiles, executable):
-        queue = multiprocessing.Queue()
-        p = multiprocessing.Process(target=_launchPyOCDGDBServer, args=(queue,))
-        p.start()
-        # wait for an 'alive' message from the server before starting gdb
-        # itself:
-        msg = None
-        while msg != 'alive':
-            try:
-                msg = queue.get(timeout=1.0)
-            except Queue.Empty as e:
-                msg = None
-                pass
-            if msg == 'dead' or not p.is_alive():
-                raise Exception('gdb server failed to start')
+        queue = Queue.Queue()
+        t = threading.Thread(target=_launchPyOCDGDBServer, args=(queue,))
+        try:
+            t.start()
+            # wait for an 'alive' message from the server before starting gdb
+            # itself:
+            msg = None
+            while msg != 'alive':
+                try:
+                    msg = queue.get(timeout=1.0)
+                except Queue.Empty as e:
+                    msg = None
+                    pass
+                if msg == 'dead' or not t.is_alive():
+                    raise Exception('gdb server failed to start')
+        except KeyboardInterrupt as e:
+            logger.error('stopped by user')
+            queue.put('kill')
+            t.join()
+            raise
         gdb_launcher(projectfiles, executable)
         queue.put('kill')
+        t.join()
     return launch_arm_gdb
 
 

